src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorDiagnosticsEndpoint.cs

```diff
@@ -0,0 +1,176 @@
+                return new RazorDiagnosticsResponse()
+                {
+                    Diagnostics = Array.Empty<Diagnostic>(),
+                    HostDocumentVersion = documentVersion
+                };
+            }
+
+            return await MapDiagnosticsAsync(
+                request,
+                filteredDiagnostics,
+                documentVersion,
+                documentSnapshot).ConfigureAwait(false);
+
+            // TODO; HTML filtering blocked on https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1257401
+            static bool CanDiagnosticBeFiltered(Diagnostic d) =>
+                (DiagnosticsToIgnore.Contains(d.Code?.String) &&
+                 d.Severity != DiagnosticSeverity.Error);
+        }
+
+        private async Task<RazorDiagnosticsResponse> MapDiagnosticsAsync(
+            RazorDiagnosticsParams request,
+            Diagnostic[] filteredDiagnostics,
+            int? documentVersion,
+            DocumentSnapshot documentSnapshot)
+        {
+            if (request.Kind != RazorLanguageKind.CSharp)
+            {
+                // All other non-C# requests map directly to where they are in the document.
+                return new RazorDiagnosticsResponse()
+                {
+                    Diagnostics = filteredDiagnostics,
+                    HostDocumentVersion = documentVersion,
+                };
+            }
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            var mappedDiagnostics = new List<Diagnostic>();
+
+            for (var i = 0; i < filteredDiagnostics.Length; i++)
+            {
+                var diagnostic = filteredDiagnostics.ElementAt(i);
+                var projectedRange = diagnostic.Range;
+
+                if (codeDocument.IsUnsupported() ||
+                    !_documentMappingService.TryMapFromProjectedDocumentRange(
+                        codeDocument, projectedRange,
+                        MappingBehavior.Inclusive,
+                        out var originalRange))
+                {
+                    // Couldn't remap the range correctly.
+                    // If this isn't an `Error` Severity Diagnostic we can discard it.
+                    if (diagnostic.Severity != DiagnosticSeverity.Error)
+                    {
+                        continue;
+                    }
+
+                    // For `Error` Severity diagnostics we still show the diagnostics to
+                    // the user, however we set the range to an undefined range to ensure
+                    // clicking on the diagnostic doesn't cause errors.
+                    originalRange = RangeExtensions.UndefinedRange;
+                }
+
+                diagnostic.Range = originalRange;
+                mappedDiagnostics.Add(diagnostic);
+            }
+
+            return new RazorDiagnosticsResponse()
+            {
+                Diagnostics = mappedDiagnostics.ToArray(),
```

Comment:
Shouldn't need to ToARray here. MappedDiagnostics is already a newly created list

Reply:
Not sure I follow. `RazorDiagnosticsResponse.Diagnostics` expects a `Diagnostics[]`. `mappedDiagnostics` was a `List<Diagnostic>` (now a `IReadOnlyList<Diagnostic>`). Wouldn't this not be required?

Reply:
List == Diagnostics[]. I think I might be missing something here

Reply:
![image](https://user-images.githubusercontent.com/14852843/103567102-30bf4d00-4e91-11eb-8b83-123bb8eed34e.png)

So wouldn't we still need a ToArray() here? Or are you suggesting something like:
![image](https://user-images.githubusercontent.com/14852843/103567186-52b8cf80-4e91-11eb-87c8-20af84d93bf0.png)

Reply:
Ah thanks for the pictures. Why is [mappedDiagnostics an IReadOnlyList now](https://github.com/dotnet/aspnetcore-tooling/pull/2832/files/298c8ccd73e68d5221f0b5a20ade26eb74020d8d#diff-4233fb067b526a5ca9d3d24db6fe013e92c3d7bb15b60b597503fb7d3a4eca4eR139)? Shouldn't just the filtered diagnostics from the parameters be an IReadOnlyList?

Reply:
mappedDiagnostics is a List as we aren't assured mappedDiagnostics.Length == filteredDiagnostics.Length (due to failed mappings).

Updated slightly https://github.com/dotnet/aspnetcore-tooling/commit/daf922a0f247e8c8fa1acbcdcd7083063ea9b36c.

Essentially, we `Add` to the mapped diagnostics the diagnostics we successfully map. We now return a diagnostic array from the `MapDiagnosticsAsync` method, however this still requires `ToArray` which was the initial concern.

Reply:
oooo I see ok ya that makes sense I forgot you were refactoring that bit